content/blog/managing-aws-credentials-on-cicd/index.md

```diff
@@ -0,0 +1,231 @@
+Your CI/CD provider likely has built-in support for handling “secrets” or “secure variables.” Typically these are user-supplied configuration values that are write-only and only accessible by the CI/CD worker jobs at runtime. For example, Travis CI has [encrypted environment variables](https://circleci.com/docs/2.0/env-vars/#overview) or [secure environment variables](https://circleci.com/docs/2.0/env-vars/#overview) in CircleCI.
+
+## Automate Rotating and Revoking those Credentials
+
+When a CI/CD system has access to an IAM User’s AWS credentials, we need to take steps to rotate those credentials regularly, even if we trust the CI/CD provider to secure them properly.
+
+> This is one of the advantages of running your CI/CD system within your AWS account. If your deployment workers are simply EC2 instances, you wouldn’t need to share IAM User credentials. The EC2 instances themselves [can be configured to assume an IAM Role](https://aws.amazon.com/blogs/security/new-attach-an-aws-iam-role-to-an-existing-amazon-ec2-instance-by-using-the-aws-cli/).
+>
+> The details about how IAM Roles are a more secure alternative to IAM User credentials are covered in the next section.
+
+“Just create a new IAM User credential, log into the CI/CD system, and then update the job’s settings” sounds simple enough. But it’s likely going to be tedious, and if you need to repeat the process for dozens if not hundreds of CI/CD pipelines, you need a better solution. Unless you start with an automated approach for key rotation, you are unlikely to add one later.
+
+The AWS security blog describes how to [rotate access keys for IAM Users](https://aws.amazon.com/blogs/security/how-to-rotate-access-keys-for-iam-users/).
+
+1. Create a second access key in addition to the one in use.
+1. Update all your applications to use the new access key and validate that the applications are working.
+1. Change the state of the previous access key to inactive.
+1. Validate that your applications are still working as expected.
+1. Delete the inactive access key.
+
+## During deployments, exchange the IAM user credentials for a more privileged IAM role
+
+Now that we have a regularly rotated, low-privileged credential to use during deployments, the next step is to use “assume role” to exchange those IAM User credentials for an IAM Role credential. This sounds confusing at first. Why exchange one credential for another? Why not just use the same credentials? In other words, what is an IAM Role, and why is this so much better than an IAM User? The credentials used to assume an IAM Role are only valid for a short period (minutes to hours). After a fixed duration, there is no threat if disclosed.
+
+Earlier, we mitigate potentially exposing IAM User credentials by using our CI/CD provider’s ability to store secret values and automatically rotating credentials. Using the IAM Role credentials adds another layer of security, but within the context of the machine itself during the deployment.
+
```

Comment:
We probably want to add a note explaining that it is the IAM Role that needs the permissions to actually deploy your Pulumi program. (e.g. access to the required APIs in your production account.)

However, we still want to be careful in the permissions granted to the IAM Role policy. For example, explicitly disallowing the `iam:*` permission is probably a good start. (Link to this [blog post](https://know.bishopfox.com/research/privilege-escalation-in-aws) that goes into the details of how having access to some IAM APIs would allow a compromised credential to fully take over an AWS account!)